### PR TITLE
PHOENIX-6960 Scan range is incorrect when query desc columns

### PR DESCRIPTION
**Pull request by virajjasani**
Jira: PHOENIX-6960

### CODE REVIEW

**Comment by virajjasani** [comment]
build results: https://ci-hadoop.apache.org/job/Phoenix/job/Phoenix-PreCommit-GitHub-PR/job/PR-1663/4/testReport/

File: phoenix-core/src/main/java/org/apache/phoenix/compile/WhereOptimizer.java [hunk 1]
```diff
@@ -1722,7 +1725,10 @@ public KeySlots visitLeave(LikeExpression node, List<KeySlots> childParts) {
                 lowerRange = Arrays.copyOf(lowerRange, lowerRange.length+1);
                 lowerRange[lowerRange.length-1] = QueryConstants.SEPARATOR_BYTE;
             }
-            KeyRange range = type.getKeyRange(lowerRange, true, upperRange, false, SortOrder.ASC);
+            boolean lowerInclusive = column.getSortOrder() == SortOrder.ASC;
+            boolean upperInclusive = column.getSortOrder() == SortOrder.DESC;
+            KeyRange range = type.getKeyRange(lowerRange, lowerInclusive, upperRange,
+                    upperInclusive, SortOrder.ASC);
```
Comment (stoty) [inline phoenix-core/src/main/java/org/apache/phoenix/compile/WhereOptimizer.java:1731]: I have debugged into this, and after inverting twice, we end up returning a non-inverted range, which eventually gets inverted in WhereOptimizer.pushKeyExpressionsToScan().

Do we even need to take SortOrder into account here, and if yes, couldn't we simplify this logic ?

Comment (stoty) [reply phoenix-core/src/main/java/org/apache/phoenix/compile/WhereOptimizer.java:1731]: I'm talking the whole method, not the specific line above.

File: phoenix-core/src/main/java/org/apache/phoenix/compile/WhereOptimizer.java [hunk 2]
```diff
@@ -1691,8 +1691,10 @@ public KeySlots visitLeave(LikeExpression node, List<KeySlots> childParts) {
             KeySlots childSlots = childParts.get(0);
             KeySlot childSlot = childSlots.getSlots().get(0);
             final String startsWith = node.getLiteralPrefix();
-            SortOrder sortOrder = node.getChildren().get(0).getSortOrder();
-            byte[] key = PVarchar.INSTANCE.toBytes(startsWith, sortOrder);
+            // TODO: is there a case where we'd need to go through the childPart to calculate the key range?
+            PColumn column = childSlot.getKeyPart().getColumn();
+            PDataType type = column.getDataType();
+            byte[] key = PVarchar.INSTANCE.toBytes(startsWith, column.getSortOrder());
```
Comment (stoty) [inline phoenix-core/src/main/java/org/apache/phoenix/compile/WhereOptimizer.java:1697]: This is where we get an inverted range

File: phoenix-core/src/main/java/org/apache/phoenix/compile/WhereOptimizer.java [hunk 3]
```diff
@@ -1722,7 +1725,10 @@ public KeySlots visitLeave(LikeExpression node, List<KeySlots> childParts) {
                 lowerRange = Arrays.copyOf(lowerRange, lowerRange.length+1);
                 lowerRange[lowerRange.length-1] = QueryConstants.SEPARATOR_BYTE;
             }
-            KeyRange range = type.getKeyRange(lowerRange, true, upperRange, false, SortOrder.ASC);
+            boolean lowerInclusive = column.getSortOrder() == SortOrder.ASC;
+            boolean upperInclusive = column.getSortOrder() == SortOrder.DESC;
+            KeyRange range = type.getKeyRange(lowerRange, lowerInclusive, upperRange,
+                    upperInclusive, SortOrder.ASC);
             if (column.getSortOrder() == SortOrder.DESC) {
                 range = range.invert();
```
Comment (stoty) [inline phoenix-core/src/main/java/org/apache/phoenix/compile/WhereOptimizer.java:1733]: And this is wehere we re-invert, getting a normal range back.

**Comment by virajjasani** [comment]
Thank you for taking a look @stoty, and yes you are correct that we have scope for optimization here, i just addressed your review with the latest revision. Also, @jinggou found some interesting test case failures, i tried to address them.

Created 5.1 backport PR: #1668

@stoty @jinggou could you please take a look again? (not urgent, i might likely be able to come online after 5 days)

**Review by stoty — APPROVED** [review]
+1 LGTM

**Comment by tkhurana** [comment]
@virajjasani can you also add some unit tests in `WhereOptimizerTest` . Those tests verify that the scan range is generated correctly. I want to make sure that we have some tests for descending columns with like expressions.

**Comment by jinggou** [comment]
LGTM. The test case that failed before passes now (e.g., descending columns with LIKE x%).

**Comment by virajjasani** [comment]
Thank you for the reviews @stoty @tkhurana @jinggou

Tanuj, sure let me add some test in `WhereOptimizerTest`, sounds good

**Comment by virajjasani** [comment]
done

**Review by tkhurana — APPROVED** [review]

File: phoenix-core/src/main/java/org/apache/phoenix/compile/WhereOptimizer.java [hunk 4]
```diff
@@ -1691,8 +1691,10 @@ public KeySlots visitLeave(LikeExpression node, List<KeySlots> childParts) {
             KeySlots childSlots = childParts.get(0);
             KeySlot childSlot = childSlots.getSlots().get(0);
             final String startsWith = node.getLiteralPrefix();
-            SortOrder sortOrder = node.getChildren().get(0).getSortOrder();
-            byte[] key = PVarchar.INSTANCE.toBytes(startsWith, sortOrder);
+            // TODO: is there a case where we'd need to go through the childPart to calculate the key range?
+            PColumn column = childSlot.getKeyPart().getColumn();
+            PDataType type = column.getDataType();
+            byte[] key = PVarchar.INSTANCE.toBytes(startsWith, SortOrder.ASC);
```
[the remark this answers is not in the document]
Comment (virajjasani) [reply phoenix-core/src/main/java/org/apache/phoenix/compile/WhereOptimizer.java:1697]: interesting! do we support `col1 LIKE ('xy%' || col2)` case? 
i tried this and saw that we return right here due to empty childParts:
```
        @Override
        public KeySlots visitLeave(LikeExpression node, List<KeySlots> childParts) {
            // TODO: optimize ILIKE by creating two ranges for the literal prefix: one with lower case, one with upper case
            if (childParts.isEmpty()) {
                return null;
            }
...
...
```

[the remark this answers is not in the document]
Comment (virajjasani) [reply phoenix-core/src/main/java/org/apache/phoenix/compile/WhereOptimizer.java:1697]: Even `col1 LIKE col2` doesn't seem to be working as childParts are coming empty, resulting in null KeySlots

[the remark this answers is not in the document]
Comment (stoty) [reply phoenix-core/src/main/java/org/apache/phoenix/compile/WhereOptimizer.java:1697]: Makes sense, we cannot use values from the table for constructing scan filters on the rowkey (unless they are coming from uncorrelated subqueries, but those are effectively constants)

[the remark this answers is not in the document]
Comment (dbwong) [reply phoenix-core/src/main/java/org/apache/phoenix/compile/WhereOptimizer.java:1697]: Thanks for the test, @virajjasani I do think that in theory `'xy%' || col2 `would be extractable as the prefix here is a constant though I was guessing we don't try to optimize this currently.

**Review by dbwong — APPROVED** [review]

File: phoenix-core/src/main/java/org/apache/phoenix/compile/WhereOptimizer.java (hunk 4 above)
[the remark this answers is not in the document]
Comment (virajjasani) [reply phoenix-core/src/main/java/org/apache/phoenix/compile/WhereOptimizer.java:1697]: Thanks for the review @dbwong

**Pull request merged**